cdap-proto/src/main/java/co/cask/cdap/proto/security/Principal.java

```diff
@@ -38,8 +38,10 @@
 
   private final String name;
   private final PrincipalType type;
-  private final int hashCode;
   private final String kerberosPrincipal;
+  // This needs to be transient because we don't want this to be populated during deserialization since that
+  // value will not be provided by the client.
+  private transient Integer hashCode;
```

Comment:
If it is a general object, then yes. Since it assign to Integer, which the `Integer.value` is a final field, so it's guaranteed that there is no partial initialization.

We can also change the `hashCode` to `int` type and do the computation if `hashCode == 0` (the same as what `String` does http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/String.java#l1452)